### DISCUSSION TITLE
Potential MEV Attacks on FundMe Contract Withdrawals

### DISCUSSION THREAD

**Discussion by padmanabhan-r** [Cyfrin/foundry-full-course-cu#828] [Q&A]
Hello everyone,

I've been diving into the "FundMe" section of the course and recently watched a video where Patrick discussed MEV attacks. This got me thinking:

If someone were to create a "FundMe" contract on the mainnet and then attempt a withdrawal, is there a risk of it being targeted by MEV attacks? If yes, what would be the recommended way to handle or mitigate this risk?

Any insights or suggestions would be greatly appreciated!

Thank you in advance.

**Comment by Niferu** [comment]
Hi @padmanabhan-r 

I have asked Patrick if protecting this function with modifiers like onlyOwner or just using separate arrays to check if there is anything to withdraw indeed for calling address will work, but he said... just look below :D

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/110537615/a198c90a-5bf8-4137-9765-cf108ca67eb1)

So this is a really good question and I would like to know also, because if we are able to code bot to outplay our protected contract (the one including a lot of restrictions like onlyOwner, checking each address balance etc) then using withdraw functions would be just stupid as we would lose money anyway. Take a look at "protected" withdraw function below. I would love to hear if it is hackable or if it is safe to use on mainnet.

```solidity
    function withdrawPending() external payable nonReentrant {
        uint256 amount = pendingReturns[msg.sender];

        if (amount > 0) {
            pendingReturns[msg.sender] = 0;
        } else {
            revert Abstract__NotEnoughETH();
        }

        (bool success, ) = msg.sender.call{value: amount}("");
        if (!success) {
            pendingReturns[msg.sender] = amount;
            revert Abstract__TransferFailed();
        }

        emit NFT_PendingBidsWithdrawal(amount, msg.sender, success);
    }
```
I believe it is safe to use as whole idea of smart contract would be just too dangerous to use if it is hackable, but let the experts talk on that @PatrickAlphaC can you kindly provide your insights? Thank you in advance!

**Reply by PatrickAlphaC** [reply]
"Hackable" is too subjective here.

- what is this function supposed to do?
- who is supposed to call it? 

"Safe" is subjective, it depends on what the protocol is supposed to do. Also, let's try to keep questions here related to the course material.

**Reply by Niferu** [reply]
Hi @PatrickAlphaC 

This function is slightly modified withdraw from course. I modified contract to allow users bidding NFT's, once there is new highest bid previous one is getting counted as balance to withdraw by previous highest bidder. As you can see everyone can call this function but only addresses with that outbidded balance are able to withdraw without getting revert `Abstract__NotEnoughETH()`.

This above is anyway only for example purposes we can also take course withdraw function, which is onlyOwner restricted, above one has more potential to get overtaken I believe thats why I recalled it instead.

```solidity
function withdraw() public onlyOwner {
        for (
            uint256 funderIndex = 0;
            funderIndex < s_funders.length;
            funderIndex++
        ) {
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // Transfer vs call vs Send
        // payable(msg.sender).transfer(address(this).balance);
        (bool success, ) = i_owner.call{value: address(this).balance}("");
        require(success);
    }
```
